fix(icon, link): improve a11y markup for screen readers

jcfranco opened the pull request:
**Related Issue:** #646 

## Summary

<!--

Please make sure the PR title and/or commit message adheres to the https://conventionalcommits.org/ specification.

Note: If your PR only has one commit and it is NOT semantic, you will need to either

a. add another commit and wait for the check to update
b. proceed to squash merge, but make sure the commit message is the same as the title.

This is because of the way GitHub handles single-commit squash merges (see https://github.com/zeke/semantic-pull-requests/issues/17)

If this is component-related, please verify that:

- [ ] feature or fix has a corresponding test
- [ ] changes have been tested with demo page in Edge

-->

This tweaks the ARIA attributes used by link and icon to improve screen reader interaction.

macandcheese commented:
Couple small things:

Previously - when a link was focused, the href would appear in the "link preview" in bottom left of browser (in Chrome at least) - now it only shows on hover (current behavior: https://esri.github.io/calcite-components/?path=/story/components-link--simple&knob-containing font size=16&knob-containing font weight=400&knob-href=http://google.com&knob-disabled=&knob-text=link text here)

Focus state no longer uses the custom "blue outline" styling - should revert that to use the existing style (and, probably, use the -outset variant vs. the current -inset).

Dev demo examples should use "theme=dark" (not something in this PR, it was always incorrect).

jcfranco commented:
Great feedback, thanks! A couple of follow-up comments:

> Previously - when a link was focused, the href would appear in the "link preview" in bottom left of browser (in Chrome at least) - now it only shows on hover (current behavior: https://esri.github.io/calcite-components/?path=/story/components-link--simple&knob-containing font size=16&knob-containing font weight=400&knob-href=http://google.com&knob-disabled=&knob-text=link text here)

I think this is a downside of this approach since focus no longer lands on the anchor. These 'fake anchor' examples suffer from the same thing:

* https://www.w3.org/TR/2016/WD-wai-aria-practices-1.1-20161214/examples/link/link.html.
* https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_link_role#:~:text=Result

Still looking for a solution.

> Focus state no longer uses the custom "blue outline" styling - should revert that to use the existing style (and, probably, use the -outset variant vs. the current -inset).

Does this mean I should drop the focus styling on the internal anchor, span elements and apply the focus-base/outset mixins to the host?

macandcheese commented:
> Does this mean I should drop the focus styling on the internal anchor, span elements and apply the focus-base/outset mixins to the host?

Yeah, I think just make sure there isn't a "browser default grey ring" around the host - If the internal elements no longer get focus themselves moving to the host makes sense.

macandcheese commented:
Would there be a way to ignore / aria-hidden on the host so screen readers just skip directly to the rendered span or anchor?

jcfranco commented:
> Would there be a way to ignore / aria-hidden on the host so screen readers just skip directly to the rendered span or anchor?

I tried that, but setting `aria-hidden="true"` on the host causes the screen reader to ignore the element. There's also some shared styles that hide any component with that attribute set.

driskull commented:
> I tried that, but setting aria-hidden="true" on the host causes the screen reader to ignore the element.

Can you set a role of presentation on the host maybe?

driskull commented on src/components/calcite-link/calcite-link.tsx:
```diff
@@ -110,7 +111,29 @@ export class CalciteLink {
+  //--------------------------------------------------------------------------
+
+  @Listen("click")
+  protected handleClick(event: KeyboardEvent | MouseEvent): void {
```
IMO it seems better to just have the click event on the childEl instead of having to capture it on the parent and then programmatically call it on the child.

Is there not a way to fix the screen reader without doing this? It just seems a little hacky :)

jcfranco replied:
This PR tries to prevent interaction with the internal anchor to prevent focusing on it after being activated. I'll see if I can make it less hacky.

paulcpederson commented:
So I just checked out this branch on NVDA and there are some links that are really weird. It is reading all of the stuff, so 

```
<calcite-link title="in the middle" href="/">in the middle</calcite-link>
```

is read:

> In the middle, visited link, in the middle, link, in the middle

Which is sort of a mess...

paulcpederson commented:
Ideally, we'd only have the interior link and that's what would be focused, read, clicked

driskull commented:
I'm not sure if the `delegatesFocus` option would help here. The bummer is it doesn't work in FF.

/**
     * When set to `true`, specifies behavior that mitigates custom element issues
     * around focusability. When a non-focusable part of the shadow DOM is clicked, the first
     * focusable part is given focus, and the shadow host is given any available `:focus` styling.
     */
    delegatesFocus?: boolean;

jcfranco commented:
> So I just checked out this branch on NVDA and there are some links that are really weird. It is reading all of the stuff, so

I see this with VoiceOver on `master`.

> Ideally, we'd only have the interior link and that's what would be focused, read, clicked

Agreed, but having an internal anchor breaks this use case `This also fixes links not getting focus after searching with the browser's find bar and closing it.` Maybe I should treat this separately.

> I'm not sure if the delegatesFocus option would help here. The bummer is it doesn't work in FF.

I did try this, but had some weird behavior. It's also not supported in Safari.

jcfranco commented:
TLDR; taking another look at this. 😂

paulcpederson commented:
Would this be a scenario where it would be better to not use shadow dom?

eriklharper commented:
> Would this be a scenario where it would be better to not use shadow dom?

`calcite-label` was converted from shadow to scoped and I think it has worked out for the better because even though you sacrifice some control over styling encapsulation, it just makes the component simpler and you don't have to hack around anything with regards to accessibility concerns.  I can see the same principle applying to anchor tags as well.  Plus, I've found that the scoped styling works pretty well as far as encapsulation goes too.

macandcheese commented:
Yeah that seems fair here.. as browsers handle shadow better down the road we can revisit those components we don’t use shadow on and see if we can add it back. Input / third parent password managers was a similar story.

Losing out on styling control is a concern but one that’s greatly outweighed by the components working, lol.

jcfranco commented:
If we ignore the additional use case I tried to fix (focusing directly on the link after find bar search is closed), the original issue can still be addressed while using shadow DOM. Scoped may work for styling, but it still exposes internals to DOM queries and high specificity selectors.

eriklharper commented:
Another option is to render a native `<a>` tag outside the shadowRoot programmatically, but I would also be cautious and make sure it works in a maquette environment too where things are unpredictable.

jcfranco commented:
> Is that when the browser in page search targets a calcite link with a partial text match? Or when a link is focused, and a user uses the in page search for anything , and then closes it?

@macandcheese Correct.

jcfranco commented:
Simplified some stuff to address the original issue:

* Used `role=presentation` to have screen reader no longer consider it a group (thanks @driskull!)
* Restored original code, so link content is focusable and we get the native browser behavior back.
* Added `aria-hidden` to icon when presentational to avoid weird scenario where its inlined `<style>` content was being read out.

> In the middle, visited link, in the middle, link, in the middle

I checked this and I think the duplicate messaging is caused by the `title` attribute set on some of the links.

driskull reviewed (approved):
🍌 💥
